main/src/ca/uwaterloo/flix/language/phase/Verifier.scala

```diff
@@ -235,6 +242,68 @@ object Verifier {
           check(expected = MonoType.Enum(sym.enumSym))(actual = t1, loc)
           tpe
 
+        case AtomicOp.ArrayLength =>
+          val List(arrt) = ts
+          arrt match {
+            case MonoType.Array(_) => check(expected = MonoType.Int32)(actual = tpe, loc)
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+
+        case AtomicOp.ArrayNew =>
+          val List(elmt, lent) = ts
+          check(expected = MonoType.Int32)(actual = lent, loc)
+          check(expected = MonoType.Array(elmt))(actual = tpe, loc)
+
+        case AtomicOp.ArrayLit =>
+          tpe match {
+            case MonoType.Array(elmt) =>
+              ts.foreach(t => check(expected = elmt)(actual = t, loc))
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+          tpe
```

Comment:
I think it looks better with tpe inside the array case, then there's no exception "in the middle"

This also goes for some following cases